src/piece_picker.cpp

```diff
@@ -1060,21 +1060,43 @@ namespace libtorrent {
 
 		piece_pos& p = m_piece_map[index];
 		int const prev_priority = p.priority(this);
-		erase_download_piece(i);
-		int const new_priority = p.priority(this);
+		if (!blocks.empty())
+		{
+			auto const binfo = mutable_blocks_for_piece(*i);
+			for (int block : blocks)
+			{
+				block_info& info = binfo[block];
+				TORRENT_ASSERT(info.state >= block_info::state_writing);
+				if (info.state == block_info::state_writing)
+					--i->writing;
+				else if (info.state == block_info::state_finished)
+					--i->finished;
```

Comment:
don't you need to check for `block_info::state_requested` and decrement `i->requested` also?

Reply:
(this should be covered by tests in `test_piece_picker.cpp`)

Reply:
This loop is resetting the state of blocks which have failed hash check so I don't think it's possible for them to be in `state_requested`, hence the assert than `info.state >= block_info::state_writing`. If it is possible for a failed block to be in `state_requested` then it should be left in that state and the request allowed to complete, hopefully the new data will turn out to be correct.